### DISCUSSION TITLE
SC as a Context Menu List

### DISCUSSION THREAD

**Discussion by FelipeRearden** [Taitava/obsidian-shellcommands#74] [Ideas]
Hello @Taitava !!!

I'd like to present a user case for an idea that I had :)

### Fundamentals
A tiny list box (context menu) that is triggered near  to the caret position when hit by the SC hotkey. 

Each item of the list `item_1,item_2, ...` is an alias of an SC defined in Settings

This image is from Chetachi's new plug-in [Highlightr](https://github.com/chetachiezikeuzor/Highlightr-Plugin) shows what I'm trying to show :)
![Screen Shot 2021-10-13 at 09 11 00](https://user-images.githubusercontent.com/85180696/137134846-1400d780-9d04-44f5-bd28-850c0a850037.png)

If is possible: 🙏
- Keyboard Navigation with arrows inside the context menu ❤️ 

### User Case Example: Action Tags/Placeholders
When reading a long note, I often use some special tags to put signals in to the note as reminders for future readings 

- #Status/🟡 to mark this note as unfinished
- #Status/🛑 to remember where I stopped 
- [[🔄]] note to mark that some concept is not 100% clear to me

My idea is to have a context menu with this three actions as SC:  

SC 1 - [alias] 🟡
`printf "#Status/🟡"` [output:Current File: Caret Position] 

SC 2 - [alias] 🛑
`printf "#Status/🛑"` [output:Current File: Caret Position] 

SC 3 - [alias] 🔄
`printf "[[🔄]]"` [output:Current File: Caret Position] 

The `input` would be a menu context near the caret position with list of three items with these alias:
 
  🟡    
  🛑    
  🔄  

---
I have one more user case that have the same meaning applied to use html tags modified by css snippets to decorate the text. The principle is the same: mini context menu near the caret postionm that let me select a list of SC to be executed :)

---

Have a great day!

**Comment by Taitava** [comment]
How about using Obsidian's context menu for this? Then it's not needed to build a completely new menu and no need to create a hotkey for opening that new menu, as Obsidian's context menu can be opened just by right clicking text.
![kuva](https://user-images.githubusercontent.com/13002142/137576674-3d5aab74-38ff-4e1c-99e1-1a8d9eda3942.png)

**Reply by FelipeRearden** [reply]
Hello @Taitava !!!!!

> How about using Obsidian's context menu for this? Then it's not needed to build a completely new menu and no need to create a hotkey for opening that new menu, as Obsidian's context menu can be opened just by right clicking text.

I have a few reasons to not use Obsidian's context menu: 


**01 Many Commands 30 for 1 context menu list**

My Obsidian usage have many different workflows (more on reason 03). This way, the Obsidian's context menu would have around 20 commands making this a not got option :)

**02 Keyboard Arrows Navigation Does not work**

I can't use the keyboard to navigate between commands in the Obsidian's context menu :(


**03 Different Commands for Different Workflows**

My idea for this new feature is to create 4/5 SC Menu lists each one for different workflows. This way I have tiny lists (2 to 5 SC commands) defined with different hotkey to quick access commands based on each workflow. 

A few user cases for me: 

A- I already showed my "remiders" tags;

B- HTML tags to highlight text in my reading workflow (I don't use <mark> tag as Chetachi does);

C- HTML tags that I have to creat pseudo headers in my writing workflow (I like to divide sections of text inside headers)

D- I use three different `<hr>`tags as horizontal line to separate text based on the type of content

E- Send selection to my two different management notes INP.md and Tasks.md (menu list to choose between these two)

F- Different types/sizes of tables templates

G- HTML tags for beautiful `<ul>` and `<ol>` pseudo lists  

Since I can use Obsidian URI with SC, I'm able to create a SC for every command form the Command Palette and then create a SC Menu List with a mix of SC commands and Command Palette commands (that is related to caret position of course) !!!!!!!!
 

And I'm only talking about Menu lists applied to caret position.
- I already have other lists using QuickAdd to open customizable modals to a subset of QuickAdd Macros. 

**03 MacOS does not have an dedicated hotkey to invoke the right click menu using the keyboard**

And the solutions I search for mimic this on MacOS are the kind of that I have to lost something else to make this work :(

I have been trying to solve this with 4 different developers for the past six months. Obsidian does not have yet this command mapped to be created a command palette command to be able to define a hotkey to this command :(

---

That's why I got so excited when I saw Chetachi's idea for <mark> . We could have this same function applied to SC !!!!! (If is feasible)


Have a great day!!!!!!

**Reply by Taitava** [reply]
> 02 Keyboard Arrows Navigation Does not work
>
> I can't use the keyboard to navigate between commands in the Obsidian's context menu :(

Is this specifically for Mac? it works on Windows at least. It's a crucial feature, so it should work on any machine. If it's not, Obsidian should fix it. Another thing is that if I would create another menu for this, separate from the context menu I suggested before, it would still use the same mechanic as Obsidian uses, as I do not see it wise to create a new style for a feature that already has an official style in Obsidian.



> **03 Different Commands for Different Workflows**
> 
> My idea for this new feature is to create 4/5 SC Menu lists each one for different workflows. This way I have tiny lists (2 to 5 SC commands) defined with different hotkey to quick access commands based on each workflow.

Ok, now this makes a bit more sense, if you are not putting all 20/30 commands into the same menu.

> 03 MacOS does not have an dedicated hotkey to invoke the right click menu using the keyboard

That's super bad, and unfortunately it's also true in many Windows keyboards nowadays, too. Maybe fifteen years ago keyboards used to have the context menu button (I don't know it's real name), but today's keyboards have replaced it with an `Fn` key. Or at least that's the case on keyboards I have used. I might switch to a freely mappable keyboard some day, preferably [a split keyboard](https://duckduckgo.com/?q=split+keyboard&t=h_&iax=images&ia=images), but that's another story.

---

I have to think and plan this still. I'd summarise it like this:
- The menu would use same style and technic as Obsidian's context menu.
- Obsidian's context menu would be one option where shell commands could be inserted into.
- Additional menus, opened up with keyboard shortcuts... I think it would need more planning about managing a mass of shell commands. Not just context menus, but command palette, too. SC settings also. I think all of those three things could benefit from grouping. I might also think what if another plugin could take care of splitting Obsidian commands to multiple command palettes and multiple context menus? Dividing context menus feels to me something that is bit drifting away from the core point of this plugin, which is to execute shell commands. This is not a no to your proposal, at least yet, but I really need to think this through first.

Have a nice day, too! 🙂

**Reply by FelipeRearden** [reply]
> Is this specifically for Mac? it works on Windows at least.

My mistake @Taitava , it's working on MacOS. I change the CSS and the select item in the list don't have a different color on hover.

Changing my CSS Snippet I could confirm that you are absolutely right: works on Mac the same way as on Windows:

![Screen Shot 2021-10-24 at 08 20 27](https://user-images.githubusercontent.com/85180696/138591955-2f0ff24b-ea4e-4507-8ea2-d0df99b4da96.png)

![Screen Shot 2021-10-24 at 08 19 59](https://user-images.githubusercontent.com/85180696/138591960-b8768a2f-d7a4-4c55-91fa-561ffe9c8c89.png)

And I can navigate (arrows keys) inside the context menu with keyboard clipboard :)

> as I do not see it wise to create a new style for a feature that already has an official style in Obsidian

100% agree 

> That's super bad, and unfortunately it's also true in many Windows keyboards nowadays, too.

Yeah. It's something that I can't believe that they don't "fix". MacOS have some tiny things that drives me crazy sometimes :(

> I might switch to a freely mappable keyboard some day, preferably a split keyboard, but that's another story

I saw some videos and its something that I wanna try one day. It don't sell in my country :( 

> I have to think and plan this still. I'd summarise it like this:
> - The menu would use same style and technic as Obsidian's context menu.
> - Obsidian's context menu would be one option where shell commands could be inserted into.

100% agree.

> I might also think what if another plugin could take care of splitting Obsidian commands to multiple command palettes and multiple context menus? 

Just as an information. QuickAdd do this for command palettes. I can group commands in a group and define a hotkey that will open only the commands of that group.

But no for context menus as a interface of selecting commands. 

> Dividing context menus feels to me something that is bit drifting away from the core point of this plugin, which is to execute shell commands.

I understand. When I wrote this FR inside the {{input}} discussion, I was thinking that a menu list was a type of UI for input (as a different modal, or a text box or multiple selection). 

In my mind was just an variation. In a what that I could type SC as a input and create a SC `execute {{input}}` which will execute the input selected in the Menu List. 

But nw I know that is much more than that :)

> This is not a no to your proposal, at least yet, but I really need to think this through first

No problem :) thanks for thinking about this in some time in the future. I can imagine that it's a feature that needs a lot of thinking and planning. We have to think about UI as well. 

---

At least I bring to you a real world user case for selecting a group/subset of SC commands. 

I wish you a fantastic Monday :)

**Reply by Taitava** [reply]
Some of your quotes had multiple times this:
> Is this specifically for Mac? it works on Windows at least.

So I wasn't able to fully follow 🙂 .

But generally about this idea, I'll keep it in my head and try to think how to proceed with it. 🌞 

Have a nice day, too! 😎

**Reply by FelipeRearden** [reply]
Sorry, yesterday was a difficult day  :) I just updated the post.

Have a great day!

**Reply by Taitava** [reply]
> 03 MacOS does not have an dedicated hotkey to invoke the right click menu using the keyboard

@FelipeRearden I just wanted to mention that [Obsidian`0.14.3`](https://forum.obsidian.md/t/obsidian-release-v0-14-3-insider-build/34806) came to the rescue: 
> - There’s now a command to open context menu for the current cursor location.

🙂 

Have a nice day! 😎

**Reply by FelipeRearden** [reply]
Hello @Taitava !!!!!!!!!!!!

Thank you very much for sharing !!!!!

The Obsidian gods finally listen to my 🙏  !!!!

Thank you very much for letting me know I almost lost this one 👀 👀 👀 👀 

Let's wait for the public release 

---

Sorry for my late reply these past weeks ... Everything is okay, I'm just busy helping my mom at home and I finally start using Obsidian Mobile and I'm cracking my head to understand the fundamentals of mobile on iPad and replicate my workflows to mobile (it has been difficult ...)

But I'm always reading all your messages and I am always here if you need me 😃 

Have a great day !!!!!!!!!!

**Reply by Taitava** [reply]
No worries, everyone has times when they can't use much time on voluntary projects. 🙂👍

You have been really helpful and supportive, thank you for it! ❤️
